### PR TITLE
adding source for protocol emulator - https://github.com/wso2/carbon-…

### PR DESCRIPTION
**Pull request by Subasinghe**

### CODE REVIEW

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/client/processors/HttpRequestInformationProcessor.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+        if (!"http".equalsIgnoreCase(scheme) && !"https".equalsIgnoreCase(scheme)) {
+            log.error("Only HTTP(S) is supported.");
+        }
+
+        ByteBuf content;
+        HttpRequest request;
+
+        if (processorContext.getRequestBuilderContext().getBody() != null) {
+
```
Comment (nandika) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/client/processors/HttpRequestInformationProcessor.java:73]: remove unwanted space

File: emulator/pom.xml [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        </dependency>
+    </dependencies>
+
+    <properties>
+        <netty.version>4.0.30.Final</netty.version>
+        <slf4j.version>1.7.5</slf4j.version>
+        <alpn-api.version>1.0.0</alpn-api.version>
+        <testng.version>6.9.8</testng.version>
+
+        <export.package>
```
Comment (lankavitharana) [inline emulator/pom.xml:125]: how is the bundle plugin configured? is it exporting/importing correct package versions?

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/AbstractProtocolEmulator.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+/*
+ * *
+ *  * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/AbstractProtocolEmulator.java]: most of the time I see licences without this extra "*" mark, I feel it's better if we follow same here.

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/contexts/AbstractThenBuilderContext.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+package org.wso2.carbon.protocol.emulator.dsl.contexts;
+
+/**
+ * AbstractThenBuilderContext
+ *
+ * @param <T> response builder context type
+ */
+public abstract class AbstractThenBuilderContext<T extends AbstractResponseBuilderContext> {
+
+    public abstract AbstractWhenBuilderContext then(T responseBuilderContext);
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/contexts/AbstractThenBuilderContext.java:33]: javadoc?

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/contexts/AbstractWhenBuilderContext.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * AbstractWhenBuilderContext
+ *
+ * @param <T> request builder context type
+ */
+public abstract class AbstractWhenBuilderContext<T extends AbstractRequestBuilderContext> {
+
+    public abstract AbstractThenBuilderContext when(T requestContextBuilder);
+
+    public abstract AbstractOperationBuilderContext operation();
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/contexts/AbstractWhenBuilderContext.java:35]: javadocs?

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/example/http/SimpleEmulatorServer.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import static org.wso2.carbon.protocol.emulator.http.server.contexts.HttpServerResponseBuilderContext.response;
+
+/**
+ * Simple emulator server
+ */
+public class SimpleEmulatorServer {
+
+    public static void main(String args[]) {
+
+        Emulator.getHttpEmulator().server().given(configure().host("127.0.0.1").port(6065).context("/users"))
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/example/http/SimpleEmulatorServer.java:35]: how about taking these variables out as constants?

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/ChannelPipelineInitializer.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+/**
+ * ChannelPipelineInitializer
+ */
+public class ChannelPipelineInitializer extends ChannelInitializer<SocketChannel> {
+    private EmulatorType emulatorType;
+    private HttpServerInformationContext serverInformationContext;
+    private HttpClientInformationContext clientInformationContext;
+    private MockServerThread[] handlers;
+    private static final Logger log = Logger.getLogger(ChannelPipelineInitializer.class);
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/ChannelPipelineInitializer.java]: best move this to the top

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/ChannelPipelineInitializer.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    public void initChannel(SocketChannel ch) {
+        if (EmulatorType.HTTP_SERVER.equals(emulatorType)) {
+
+            initializeHttpServerChannel(ch);
+        } else if (EmulatorType.HTTP_CLIENT.equals(emulatorType)) {
+            initializeHttpClientChannel(ch);
+        }
+    }
+
+    private void initializeHttpServerChannel(SocketChannel ch) {
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/ChannelPipelineInitializer.java:69]: javadocs?

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/client/HttpClientInitializer.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    private HttpClientInformationContext clientInformationContext;
+    private EventLoopGroup group;
+    private Bootstrap bootstrap;
+
+    public HttpClientInitializer(HttpClientInformationContext clientInformationContext) {
+        this.clientInformationContext = clientInformationContext;
+
+    }
+
+    public void initialize() throws Exception {
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/client/HttpClientInitializer.java:54]: some javadocs?

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/client/HttpClientInitializer.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+ * Http client initializer
+ */
+public class HttpClientInitializer {
+    private HttpClientInformationContext clientInformationContext;
+    private EventLoopGroup group;
+    private Bootstrap bootstrap;
+
+    public HttpClientInitializer(HttpClientInformationContext clientInformationContext) {
+        this.clientInformationContext = clientInformationContext;
+
```
Comment (lankavitharana) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/http/client/HttpClientInitializer.java]: remove unnecessary space

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/example/http/SimpleEmulatorServer.java (hunk 6 above)
Comment (Subasinghe) [reply emulator/src/main/java/org/wso2/carbon/protocol/emulator/example/http/SimpleEmulatorServer.java:35]: These variables are used only for creating sample emulator server and client. They are not used in other places.

File: emulator/src/test/resources/testng.xml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+
```
Comment (hasithaa) [inline emulator/src/test/resources/testng.xml:16]: Missing License header

File: emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/contexts/AbstractEmulatorBuilderContext.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.protocol.emulator.dsl.contexts;
+
+/**
+ * Abstract class for EmulatorBuilderContext.
```
Comment (vinok88) [inline emulator/src/main/java/org/wso2/carbon/protocol/emulator/dsl/contexts/AbstractEmulatorBuilderContext.java]: Is this a abstract class??

File: emulator/src/test/resources/testng.xml (hunk 11 above)
Comment (Subasinghe) [reply emulator/src/test/resources/testng.xml:16]: already fixed.

**Pull request merged**